File: cpp/software/OperatingSystem.cpp
#pragma once
#include <bits/stdc++.h>

#include "../component/Software.cpp"
#include "../interfaces/Bootable.cpp"
#include "../interfaces/Executable.cpp"
#include "../interfaces/Upgradable.cpp"

class OperatingSystem : public Software, public Bootable, public Upgradable
{
private:
    std::string osName;
    std::string kernelVersion;
    bool is64Bit;
    bool installed;

public:
    // Constructor
    OperatingSystem() : Software("", "", 0.0), osName(""), kernelVersion(""), is64Bit(false) {}
    OperatingSystem(std::string name, std::string version, bool bit64, std::string softwareVersion, std::string licenseKey, double sizeMB)
        : Software(softwareVersion, licenseKey, sizeMB), osName(name), kernelVersion(version), is64Bit(bit64) {}

    // Destructor
    ~OperatingSystem() {}

    // Setters
    void setOsName(const std::string &name)
    {
        osName = name;
    }

    void setKernelVersion(const std::string &version)
    {
        kernelVersion = version;
    }

    void setIs64Bit(bool bit64)
    {
        is64Bit = bit64;
    }

    void setInstalled(bool isInstalled)
    {
        installed = isInstalled;
    }

    // Getters
    std::string getOsName() const
    {
        return osName;
    }

    std::string getKernelVersion() const
    {
        return kernelVersion;
    }

    bool getIs64Bit() const
    {
        return is64Bit;
    }

    bool isInstalled() const
    {
        return installed;
    }

    bool boot() override
    {
        printf("Operating System booted\n");
        return true;
    }

    void shutdown() override
    {
        printf("Operating System shutdown\n");
    }

    void restart() override
    {
        printf("Operating System restarted\n");
    }

    bool upgrade() override
    {
        printf("Operating System upgraded\n");
        return true;
    }

    bool downgrade() override
    {
        printf("Operating System downgraded\n");
        return true;
    }
};

File: python/software/operating_system.py
from components.software import Software
from interfaces.bootable import Bootable
from interfaces.executable import Executable
from interfaces.upgradable import Upgradable

class OperatingSystem(Software, Bootable, Upgradable):
    def __init__(self, os_name="", kernel_version="", is_64_bit=False, software_version="", license_key="", size_mb=0.0):
        super().__init__(software_version, license_key, size_mb)
        self._os_name = os_name
        self._kernel_version = kernel_version
        self._is_64_bit = is_64_bit
        self._installed = False

    # Setters
    def set_os_name(self, os_name):
        self._os_name = os_name

    def set_kernel_version(self, kernel_version):
        self._kernel_version = kernel_version

    def set_is_64_bit(self, is_64_bit):
        self._is_64_bit = is_64_bit

    def set_installed(self, installed):
        self._installed = installed

    # Getters
    def get_os_name(self):
        return self._os_name

    def get_kernel_version(self):
        return self._kernel_version

    def get_is_64_bit(self):
        return self._is_64_bit

    def is_installed(self):
        return self._installed

    def boot(self):
        print("Operating System booted")
        return True

    def shutdown(self):
        print("Operating System shutdown")

    def restart(self):
        print("Operating System restarted")

    def upgrade(self):
        print("Operating System upgraded")
        return True

    def downgrade(self):
        print("Operating System downgraded")
        return True